Refactor makefile and start development documentation

thetechnick opened the pull request:
Not finished yet - but I want Prow to execute tests!

Based on work put forward in https://github.com/openshift/addon-operator/pull/39

Signed-off-by: Nico Schieder <nschieder@redhat.com>

erdii commented:
/hold

erdii reviewed (approved):
lgtm until now!

sugarraysam commented:
Just reviewed the changes. I think there are still a bunch of Makefile targets I am not sure we should put in there. For example, everything related to a kind cluster could be the user responsibility. Unless of course we depend on this for CI??

I am still unclear why we want binaries to live in a complicated path like `.cache/dependencies/$(UNAME_OS)/$(UNAME_ARCH)` ?

I like the cleaner awk command this is really nice ;)

I think we could refactor getting go dependencies to a function (but I don't feel strongly about this, just more DRY).

There are still alot of targets, not clear to me if I want to run e2e test if I should `test-e2e`, `test-e2e-short` or `test-e2e-local`. Maybe only document the make targets that are meant to be used by end users? For example, we will never run `make golangci-lint` or `make opm` on its own,  but really just run `make dev-setup`.

Thanks for the good work @thetechnick !

sugarraysam commented:
(I cant review a line without change) -> but if we want to enforce versions on tooling for this project, let's also control `GOBIN`:

```make
# replace ?=  with :=
export GOBIN := $(abspath .cache/dependencies/bin)
```
I always set my `GOBIN` so right now projects override my global deps in `~/.go/bin/`. I could set GOBIN in `.envrc` but I feel we are just trying to replicate the virtual env pattern so let's do it fully.

sugarraysam commented on Makefile:
```diff
@@ -114,7 +150,10 @@ $(YQ):
 
-# setup goimports
+## Setup yq.
+yq:
```
missing `yq: $(YQ)`

sugarraysam commented on Makefile:
```diff
@@ -100,7 +133,10 @@ $(CONTROLLER_GEN):
 
-# setup yq
+## Setup controller-gen.
+controller-gen:
```
missing `controller-gen: $(CONTROLLER_GEN)`

erdii commented:
Can the go-gettable dependencies be refactored to use the `go-get-tool` idea from @sugarraysam?
I'd love to have dep declarations concise - because the real information in these declarations is not *how* they are installed but *what*  and *what version*.

thetechnick commented:
> Just reviewed the changes. I think there are still a bunch of Makefile targets I am not sure we should put in there. For example, everything related to a kind cluster could be the user responsibility. Unless of course we depend on this for CI??
I would like to execute this in CI, but we don't have an environment for that now, because OpenShift CI does not support running kind and the workarounds where a bit too much for us to tackle.

Currently creating the kind cluster is done in Makefile, just because of convinience - a few commands need a kind cluster to target and it's easy to standardize on a certain kind-cluster-name, so you don't need to pipe arguments to the Makefile.

> I always set my GOBIN so right now projects override my global deps in ~/.go/bin/. I could set GOBIN in .envrc but I feel we are just trying to replicate the virtual env pattern so let's do it fully.

I have removed using GOBIN directly and now make sure everything lands in `.cache/dependencies/bin`, if someone needs to override binaries he would have to symlink them into that directory now.

> Finally, I would also use absolute path with these tools, because right now it depends on path resolution which could be unstable?

Top of the Makefile is adding `.cache/dependencies/bin` in front of any existing `$PATH`, so bash should search there first every time. We can also do it with variables, but it definitely does not help readability.

> Maybe only document the make targets that are meant to be used by end users? For example, we will never run make golangci-lint or make opm on its own, but really just run make dev-setup.

I have removed all the individual dependency targets from `make help` to make it a bit cleaner.
I would hope that the Development documentation (Currently patched into the Readme.md) would make it easy to contribute by pointing out workflows and make commands "to get shit done".

sugarraysam commented:
@thetechnick Great work thanks for adjusting and refactoring the Makefile. The developer experience is much improved from those changes.

Went through the docs and was able to follow the `getting started`.

**make test-e2e-short**

I get a few error logs, I'm guessing this is intended as we are testing state and the Reconcile loop is allowed to error during test? The overall command exits with 0:

```bash
2021-06-22T18:12:56.007Z        ERROR   controller-runtime.manager.controller.addon     Reconciler error        {"reconciler group": "addons.managed.openshift.io", "reconciler kind": "Addon", "name": "addon-oisafbo12", "namespace": "", "error": "failed to ensure CatalogSource: Operation cannot be fulfilled on addons.addons.managed.openshift.io \"addon-oisafbo12\": the object has been modified; please apply your changes to the latest version and try again"}
github.com/go-logr/zapr.(*zapLogger).Error
<...>
2021-06-22T18:13:06.609Z        ERROR   controller-runtime.manager.controller.addon     Reconciler error        {"reconciler group": "addons.managed.openshift.io", "reconciler kind": "Addon", "name": "addon-oisafbo12", "namespace": "", "error": "failed to ensure wanted Namespaces: Operation cannot be fulfilled on addons.addons.managed.openshift.io \"addon-oisafbo12\": the object has been modified; please apply your changes to the latest version and try again"}
github.com/go-logr/zapr.(*zapLogger).Error
```

sugarraysam commented on Makefile:
```diff
@@ -26,127 +26,123 @@ UNAME_OS:=$(shell uname -s)
+.PHONY: version
+
+## Cleans cached binaries, dependencies and container image tars.
+clean:
```
Can we also add `delete-kind-cluster` as a dependency target?

thetechnick replied:
Adding.

I was 120% sure that `kind delete cluster` will error when the cluster does not exist - it does not :)

sugarraysam commented on README.md:
```diff
@@ -16,11 +16,91 @@ Addon Operator coordinates the lifecycle of Addons in managed OpenShift.
+
+**Warning:**
+- Your code runs as `cluster-admin`, you might run into permission errors when running in-cluster.
+- Code-Generators need to be re-run and CRDs re-applied via `make setup-addon-operator-crds` when code under `./apis` is changed.
```
I encountered this issue when using the `getting started docs`. Maybe add this extra section?

```markdown
## Troubleshooting

**[Set `nf_conntrack_max`](https://github.com/kubernetes-sigs/kind/issues/2240)**

When using docker to spin a new Kind cluster, `kube-proxy` would not start throwing this error:

```bash
I0511 11:47:28.965997       1 conntrack.go:100] Set sysctl 'net/netfilter/nf_conntrack_max' to XXXXXX
F0511 11:47:28.966114       1 server.go:495] open /proc/sys/net/netfilter/nf_conntrack_max: permission denied
```

Make sure to `sudo sysctl net/netfilter/nf_conntrack_max=<value>`, and add a drop-in file to `/etc/sysctl.d/99-custom.conf` to set the kernel params permanently.
```

thetechnick replied:
Added this section, very nice!

sugarraysam reviewed (approved):
Good work!

thetechnick commented:
/unhold
